rfcs/0013-configurable-and-composable-storage.md

```diff
@@ -0,0 +1,600 @@
+let value = storage.vec_of_vec_of_u64.deep_read();
+
+// In this case that type is a slice of slices of `u64`.
+assert_eq(value, [[1, 2, 3], [11, 22, 33]]);
+```
+
+For more examples on using storage types in code, see the `demo` functions in the [demo contracts](../files/0013-configurable-and-composable-storage/contracts/).
+
+## Storage API design guidelines
+
+[storage-api-design-guidelines]: #storage-api-design-guidelines
+
+Storage API design guidelines are discussed and explained in detail in sample implementations given in the [`api-design`](../files/0013-configurable-and-composable-storage/api-design/). Here we are listing the major decisions and why they make the usage of the API easier by clearly communicating potential nuances in the usage of a particular storage type.
+
+1. The API clearly communicates that the underlying storage might be uninitialized which can cause certain operations, like e.g. `StorageBox::read()` to fail. _All_ operations that can fail must have the `try_` prefix and return an `Option`.
+1. Every `try_` operation has its "non-try" counterpart that returns the value and reverts if the operation fails internally. The counterpart is always named the same as the `try_`, but without the `try_` prefix. E.g., `read()` for `try_read()`.
+1. Uninitialized storage instances are always safe to use. This way we benefit of the possibility to not pay for default initialization and still safely use the storage instance. Every storage type will provide its own semantics for uninitialized storage state. E.g. an uninitialized [`StorageVec`](../files/0013-configurable-and-composable-storage/api-design/storage_vec.sw) has the same semantics as an empty `StorageVec`.
+1. Thus, the failing of `try_` methods can always be seen as a semantic failure. E.g., `StorageVec::try_get()` can fail if the `StorageVec` is not initialized or because it requested element index is out of bounds. Since in the first case we consider the vector to be empty, it is also a semantic error.
+1. Certain operations might have a version with a "special behavior". E.g., in the case of the `StorageVec`, we want to provide a possibility to `get()` an element without the expensive length check. In such cases, the method will be named by the original name, and the suffix indicating the "special behavior". E.g., `StorageVec::get_unchecked()`.
+1. When the special behavior consists of doing deep clear via `DeepClearStorage` or deep read via `DeepReadStorage`, the suffix will always be `_deep_clear` and `_deep_read`, respectively.
+
+Thus, every storage type operation might come in all or some of these flavours:
+- `fn <operation>() -> T`: reverts in case of uninitialized storage or other error.
+- `fn try_<operation>() -> Option<T>`
+- `fn <operation>_<special_behavior>() -> T`: reverts in case of uninitialized storage or other error.
+- `fn try_<operation>_<special_behavior>() -> Option<T>`
```

Comment:
We discussed changing the naming in the API previously here: https://github.com/FuelLabs/sway/issues/4740

I think we should go with the opposite of what's said in this paragraph.

Most of the time, you want an operation to return `Option`. I'd much rather we standardize on `read`+`read_unchecked` than `try_read` + `read`. Reverts should be the exception not the rule and it should be clear when an operation potentially reverts.

Reply:
# TL; DR

The proposal (for the next discussion round :smile:) is:
- `fn <operation>() -> Option<T>`: `None` in case of uninitialized storage or other, semantic, error.
- `fn <operation>_or_revert() -> T`
- `fn <operation>_<special_behavior>() -> Option<T>`: `None` in case of uninitialized storage or other, semantic, error.
- `fn <operation>_<special_behavior>_or_revert() -> T`
- `fn <operation>() -> T`: operation never fails.
- `fn <operation>_<special_behavior>() -> T`: operation with special behavior never fails.

Looking at the storage types we have, there will never be an explosion of those variants per operation, and we still have a clear guidelines that capture the overall complexity (explained below).

# Detailed Analysis

After spending a good portion of the last week walking in circles and pondering how to capture the complexity of the storage access semantics into consistent API guidelines, here I am again, doing the same :smile:

I am completely fine with promoting the `Option` returning operations to be the default ones.

But we still need to capture the overall matrix of possible combinations. Since [FuelLabs/sway/#4740](https://github.com/FuelLabs/sway/issues/4740) is not going into details there, let's do it here.

Before starting, just to emphasise the difference, the term "unchecked" in the #4740 and in the RFC paragraph, means two different things. In the #4740 it is a synonim for unwrap while in the RFC it represents a "special behavior". In the concrete example, running the `StorageVec::get()` without checking the vector length and out-of-bounds access.

To the matrix of possible combinations and what we want to capture in the API.

Every operation can potentially:
- fail if the storage is not initialized. E.g., `StorageBox::read()` on uninitialized `StorageBox`.
- fail because of a semantically wrong usage. E.g., out of bounds `StorageVec::get()` on initialized `StorageVec`.
- have one or more "special behaviors" that affect what the operation do. E.g., `StorageVec::get()` without lenght checks. E.g., `StorageVec::pop()` with deep clear.
  - we expect a small number of special behaviors per operation, very often none or mostly one.
  - special behaviors might require additional trait constraints. E.g. `DeepClearStorage`.
  - special behavior operations can also fail because of the first two points.
- have none of the above; neither it can fail nor it has any special behaviors. E.g., `StorageMap::get()`.

The guidelines should reflect the above matrix.

What RFC proposes, to simplify the matrix and the API, and assuming that we are not losing any valuable information there, is to treat the access to the uninitialized storage as a semantically wrong usage and not as a special technical error.

Means, every storage type assigns a meaningful default semantics to the uninitialized state. E.g, an uninitialized `StorageVec` behaves the same as an empty `StorageVec`. For all the storage types we currently support, this is possible in an intuitive way.

It also has a benefit of just creating an instance with `new()` without the need for initialization. `StorageVec::new()` behaves the same as `StorageVec::init([])` but does not produce any storage writes until used. For storage types I've played with and the sample `StoragePair` having this meaningful default behavior was always possible.

If we do not join those two points and consistently want to distinguish between the failing because of the storage being uninitialized and because of a semantically wrong usage, we will necessarily end up in the cumbersome API like, e.g., `StorageVec::get()` returning `Option<Option<T>>` where the first option means failure on uninitialized and the second out-of-bounds. I doubt that developers will ever want to distinguish those two cases.

One more thing to consider in the guidelines is:
> it should be clear when an operation potentially reverts

How to achieve this via naming convention only? (We are talking here about the computational effects. We had this discussion internally triggered by the `#[storage]` attributes not being a part of the function type.)

One possibility would be:
- `<operation>() -> Option<T>`: does not revert.
- `<operation>_<unchecked or some better name>() -> T`: potentially reverts.
- `<operation>() -> T`: does not revert. E.g., `StorageMap::get() -> T`.

And all this should also be aplicable to "special behavior" variants.

Using the term "unchecked" as a suffix I see as misleading. Because, in the case of, e.g., `StorageBox::read_unchecked()` it actually _does check_ and it _will revert if that check fails_.

"unchecked" as a word implies that we are actually skipping checks and who knows what can happen. Developers responsibility. E.g., in the RFC, the `StorageVec::get_unchecked()` cannot fail because it literally _does not check_ the bounds and returns the calculated element. Accessing that element later on can fail, but the `get_unchecked()` itself cannot.

I am arguing here to adopt some other terminology for the `<operation>_<give me the result or revert if it is not there>()`, and not not to have that variant.

Perhaps `<operation>_or_revert()`?

To wrap it up, the proposal (for the next discussion round :smile:) is:
- `fn <operation>() -> Option<T>`: `None` in case of uninitialized storage or other, semantic, error.
- `fn <operation>_or_revert() -> T`
- `fn <operation>_<special_behavior>() -> Option<T>`: `None` in case of uninitialized storage or other, semantic, error.
- `fn <operation>_<special_behavior>_or_revert() -> T`
- `fn <operation>() -> T`: operation never fails.
- `fn <operation>_<special_behavior>() -> T`: operation with special behavior never fails.

Looking at the storage types we have, there will never be an explosion of those variants per operation, and we still have a clear guidelines that capture the overall complexity.

Reply:
A question that immediately comes to my mind is why `_or_revert()/unchecked()` at all, when we can simply call `unwrap()`?

The `<operation>() <-> try_<operation>()` pairs are there from the beginning if that's a valid reason. `<operation>().unwrap()` results in a potentially much bigger code then `<operation>_or_revert()` (see [FuelLabs/sway/#5982](https://github.com/FuelLabs/sway/issues/5982)), but this should be handled in optimizations.

Also, other APIs, aside from the storage API, does not have such pairs. E.g. `Vec` has only `get() -> Option<T>`.

So why to have `_or_revert()/unchecked()` so prominently in the storage API?